### PR TITLE
[RTM] Improve the locale handling

### PR DESCRIPTION
**Pull request by leofeyer**
This PR caches the installed languages in the DI container, so we do not need to iterate over the file system upon every request. Also, it no longer stores the locale in the session by default, which is now done in the `setUserFromDb()` method.

### CODE REVIEW

File: src/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -168,4 +181,41 @@ private function resolvePath($value)
...
+        $dirs = [__DIR__.'/../Resources/contao/languages'];
+
+        // app/Resources/contao/languages
+        if (is_dir($this->rootDir.'/Resources/contao/languages')) {
+            $dirs[] = $this->rootDir.'/Resources/contao/languages';
+        }
+
+        $finder = Finder::create()->directories()->depth(0)->in($dirs);
+
+        $languages = array_values(
```
Comment (aschempp) [inline src/DependencyInjection/Configuration.php]:
```php
$languages = array_map(
    function (SplFileInfo $file) {
        if (!preg_match('/^[a-z]{2}(_[A-Z]{2})?$/', $file->getFilename())) {
            return false;
        }

        return $file->getFilename();
    },
    iterator_to_array($finder)
);

array_unshift($languages, $this->defaultLocale);

return array_unique(array_values($languages));
```

Comment (aschempp) [reply src/DependencyInjection/Configuration.php]:
```php
$languages = [$this->defaultLocale];

foreach ($finder as $file) {
    if (preg_match('/^[a-z]{2}(_[A-Z]{2})?$/', $file->getFilename())) {
        $languages[] = $file->getFilename();
    }
}

return array_values(array_unique($languages));
```

Comment (leofeyer) [reply src/DependencyInjection/Configuration.php]: Changed in 6bcce6737acb8a606ab2c1f69d671e75e97e771a.

File: src/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -168,4 +181,32 @@ private function resolvePath($value)
...
+            $dirs[] = $this->rootDir.'/Resources/contao/languages';
+        }
+
+        // The default locale must be the first supported language (see contao/core#6533)
+        $languages = [$this->defaultLocale];
+
+        $finder = Finder::create()->directories()->depth(0)->in($dirs);
+
+        foreach ($finder as $file) {
+            if (preg_match('/^[a-z]{2}(_[A-Z]{2})?$/', $file->getFilename())) {
```
Comment (Toflar) [inline src/DependencyInjection/Configuration.php]: Can't the `$finder` do that already?  Like so:

```php
$finder = Finder::create()->directories()->depth(0)->in($dirs)->name('/^[a-z]{2}(_[A-Z]{2})?$/');
```

Comment (leofeyer) [reply src/DependencyInjection/Configuration.php]: Changed in 4f3558916a72d4a99603b81c3223c9fd82b621e2.

File: src/DependencyInjection/ContaoCoreExtension.php [hunk 3]
```diff
@@ -51,7 +51,8 @@ public function getConfiguration(array $config, ContainerBuilder $container)
 
         return new Configuration(
             $container->getParameter('kernel.debug'),
-            $container->getParameter('kernel.project_dir')
+            $container->getParameter('kernel.project_dir'),
```
Comment (aschempp) [inline src/DependencyInjection/ContaoCoreExtension.php]: This should be `kernel.root_dir`

File: src/Resources/contao/classes/FrontendUser.php [hunk 4]
```diff
@@ -270,9 +270,19 @@ protected function setUserFromDb()
...
+		// Set the language
+		if ($this->language)
 		{
+			if (\System::getContainer()->has('session'))
+			{
+				$session = \System::getContainer()->get('session');
+
+				if ($session->isStarted())
+				{
+					$session->set('_locale', $this->language);
```
Comment (aschempp) [inline src/Resources/contao/classes/FrontendUser.php]: This is strange, let's discuss.

File: src/DependencyInjection/ContaoCoreExtension.php (hunk 3 above)
Comment (leofeyer) [reply src/DependencyInjection/ContaoCoreExtension.php]: Changed in be52c2c8b0b316fc47ee956df5fa1013f26a1b7b.

**Comment by leofeyer** [comment]
Merged in b40a9357b775380e185888f2c19fb47060192c95.

**Pull request closed without merging**